### ISSUE TITLE
ApiGateway doesn't honor api-key-required flag on methods

### ISSUE THREAD

**Issue by marksteele** [localstack/localstack#2488]
# Type of request: This is a ...

[x ] bug report
[ ] feature request

# Detailed description

When creating a method that requires an api key, it is not enforced by the mock service.
...

## Expected behavior

Only valid api keys allowed through
...

## Actual behavior

No checking of api keys.
...

# Steps to reproduce

Create an APIG with method that requires API keys. 

## Command used to start LocalStack
```
PORT_WEB_UI=8081 SERVICES=lambda,dynamodb,apigateway localstack start --docker
```
...

## Client code (AWS SDK code snippet, or sequence of "awslocal" commands)

Create lambda.js:
```
'use strict'

const apiHandler = (payload, context, callback) => {
    console.log(`Function apiHandler called with payload ${JSON.stringify(payload)}`);
    callback(null, {
        statusCode: 201,
        body: JSON.stringify({
            message: 'Hello World'
        }),
        headers: {
            'X-Custom-Header': 'ASDF'
        }
    }); 
}
    
module.exports = {
    apiHandler,
}
```

Zip it up

```
zip api-handler.zip index.js
```

Run this:

```
#!/bin/sh

API_NAME=test
REGION=us-east-1
STAGE=prod

function fail() {
    echo $2
    exit $1
}

awslocal lambda create-function \
    --region ${REGION} \
    --function-name ${API_NAME} \
    --runtime nodejs8.10 \
    --handler lambda.apiHandler \
    --memory-size 128 \
    --zip-file fileb://api-handler.zip \
    --role arn:aws:iam::123456:role/irrelevant

[ $? == 0 ] || fail 1 "Failed: AWS / lambda / create-function"

LAMBDA_ARN=$(awslocal lambda list-functions --query "Functions[?FunctionName==\`${API_NAME}\`].FunctionArn" --output text --region ${REGION})

awslocal apigateway create-rest-api \
    --region ${REGION} \
    --name ${API_NAME}

[ $? == 0 ] || fail 2 "Failed: AWS / apigateway / create-rest-api"

API_ID=$(awslocal apigateway get-rest-apis --query "items[?name==\`${API_NAME}\`].id" --output text --region ${REGION})
PARENT_RESOURCE_ID=$(awslocal apigateway get-resources --rest-api-id ${API_ID} --query 'items[?path==`/`].id' --output text --region ${REGION})

awslocal apigateway create-usage-plan --name "basic" --description "basic" --quota limit=2000000,offset=0,period=MONTH --api-stages apiId=$API_ID,stage=prod
awslocal apigateway create-usage-plan --name "premium" --description "premium" --quota limit=5000000,offset=0,period=MONTH --api-stages apiId=$API_ID,stage=prod

awslocal apigateway create-resource \
    --region ${REGION} \
    --rest-api-id ${API_ID} \
    --parent-id ${PARENT_RESOURCE_ID} \
    --path-part "{somethingId}"

[ $? == 0 ] || fail 3 "Failed: AWS / apigateway / create-resource"

RESOURCE_ID=$(awslocal apigateway get-resources --rest-api-id ${API_ID} --query 'items[?path==`/{somethingId}`].id' --output text --region ${REGION})

awslocal apigateway put-method \
    --region ${REGION} \
    --rest-api-id ${API_ID} \
    --resource-id ${RESOURCE_ID} \
    --http-method GET \
    --request-parameters "method.request.path.somethingId=true" \
    --authorization-type "NONE" \
    --api-key-required \

[ $? == 0 ] || fail 4 "Failed: AWS / apigateway / put-method"

awslocal apigateway put-integration \
    --region ${REGION} \
    --rest-api-id ${API_ID} \
    --resource-id ${RESOURCE_ID} \
    --http-method GET \
    --type AWS_PROXY \
    --integration-http-method POST \
    --uri arn:aws:apigateway:${REGION}:lambda:path/2015-03-31/functions/${LAMBDA_ARN}/invocations \
    --passthrough-behavior WHEN_NO_MATCH \

[ $? == 0 ] || fail 5 "Failed: AWS / apigateway / put-integration"


awslocal apigateway put-method \
    --region ${REGION} \
    --rest-api-id ${API_ID} \
    --resource-id ${RESOURCE_ID} \
    --http-method POST \
    --request-parameters "method.request.path.somethingId=true" \
    --authorization-type "NONE" \
    --api-key-required \

[ $? == 0 ] || fail 4 "Failed: AWS / apigateway / put-method"

awslocal apigateway put-integration \
    --region ${REGION} \
    --rest-api-id ${API_ID} \
    --resource-id ${RESOURCE_ID} \
    --http-method POST \
    --type AWS_PROXY \
    --integration-http-method POST \
    --uri arn:aws:apigateway:${REGION}:lambda:path/2015-03-31/functions/${LAMBDA_ARN}/invocations \
    --passthrough-behavior WHEN_NO_MATCH \

[ $? == 0 ] || fail 5 "Failed: AWS / apigateway / put-integration"

awslocal apigateway create-deployment \
    --region ${REGION} \
    --rest-api-id ${API_ID} \
    --stage-name ${STAGE} \

[ $? == 0 ] || fail 6 "Failed: AWS / apigateway / create-deployment"

ENDPOINT=http://localhost:4567/restapis/${API_ID}/${STAGE}/_user_request_/HowMuchIsTheFish

echo "API available at: ${ENDPOINT}"

echo "Testing GET:"
curl -i ${ENDPOINT}

echo "Testing POST:"
curl -iX POST ${ENDPOINT}
```
...

**Comment by whummer** [comment]
Thanks for the detailed report @marksteele . Initial support for api-key-required has been added in #2520 . Note that the current implementation does not check the actual value of the passed API key - this will be added in a follow-up PR soon. Thanks

**Comment by marksteele** [comment]
Not sure if the fix works. I've built localstack from master, and tried...

```
MarkBook-Pro:fixtures marksteele$ awslocal apigateway get-method --rest-api-id 91pojyn9al --resource-id 6aegsqqqss --http-method GET
{
    "apiKeyRequired": true, 
    "httpMethod": "GET", 
    "authorizationType": "NONE", 
    "methodIntegration": {
        "integrationResponses": {
            "200": {
                "responseTemplates": {
                    "application/json": null
                }, 
                "statusCode": 200
            }
        }, 
        "passthroughBehavior": "WHEN_NO_MATCH", 
        "cacheKeyParameters": [], 
        "uri": "arn:aws:apigateway:us-east-1:lambda:path/2015-03-31/functions/arn:aws:lambda:us-east-1:000000000000:function:sample/invocations", 
        "httpMethod": "POST", 
        "cacheNamespace": "69d71ee2", 
        "type": "AWS_PROXY"
    }
}
MarkBook-Pro:fixtures marksteele$ curl http://localhost:4567/restapis/91pojyn9al/prod/_user_request_/HowMuchIsTheFish -v
*   Trying ::1...
* TCP_NODELAY set
* Connected to localhost (::1) port 4567 (#0)
> GET /restapis/91pojyn9al/prod/_user_request_/HowMuchIsTheFish HTTP/1.1
> Host: localhost:4567
> User-Agent: curl/7.64.1
> Accept: */*
> 
< HTTP/1.1 201 Created
< Server: BaseHTTP/0.6 Python/3.8.2
< Date: Sat, 06 Jun 2020 19:38:55 GMT
< X-Custom-Header: ASDF
< Content-Length: 25
< Access-Control-Allow-Origin: *
< Access-Control-Allow-Methods: HEAD,GET,PUT,POST,DELETE,OPTIONS,PATCH
< Access-Control-Allow-Headers: authorization,content-type,content-md5,cache-control,x-amz-content-sha256,x-amz-date,x-amz-security-token,x-amz-user-agent,x-amz-target,x-amz-acl,x-amz-version-id,x-localstack-target,x-amz-tagging
< Access-Control-Expose-Headers: x-amz-version-id
< 
```

Added this to the code:
```
def is_api_key_valid(is_api_key_required, headers):
    LOGGER.info('Checking for valid api key')
    if not is_api_key_required:
        LOGGER.info('No API key required')
        return True

    LOGGER.info('API KEY REQUIRED')

    if not headers.get('X-API-Key'):
        LOGGER.info('API KEY NOT FOUND')
        return False

    LOGGER.info('API KEY OK')
    # TODO: Currently, we're only checking the presence of the API key in the header.
    #  In the future we need to verify the actual key value against the REST API configuration as well.
    return True


```

Logs are showing:
```
localstack_1  | 2020-06-06T19:52:00:INFO:localstack.services.apigateway.apigateway_listener: Checking for valid api key
localstack_1  | 2020-06-06T19:52:00:INFO:localstack.services.apigateway.apigateway_listener: No API key required
```

**Comment by whummer** [comment]
Thanks for the follow-up @marksteele . Another enhancement (and a small integration test) has been pushed in #2524 - can you please give it another try with the `latest` version of the Docker image? Thanks

**Comment by marksteele** [comment]
Still doesn't appear to work. 

Pulled down master and created a new docker image.

```
echo $RESOURCE_ID $API_ID; 
oztjferq0v 4u6mm1uzzs

awslocal apigateway get-method --rest-api-id $API_ID --resource-id $RESOURCE_ID --http-method GET
{
    "apiKeyRequired": true, 
    "httpMethod": "GET", 
    "authorizationType": "NONE", 
    "methodIntegration": {
        "integrationResponses": {
            "200": {
                "responseTemplates": {
                    "application/json": null
                }, 
                "statusCode": 200
            }
        }, 
        "passthroughBehavior": "WHEN_NO_MATCH", 
        "cacheKeyParameters": [], 
        "uri": "arn:aws:apigateway:us-east-1:lambda:path/2015-03-31/functions/arn:aws:lambda:us-east-1:000000000000:function:sample/invocations", 
        "httpMethod": "POST", 
        "cacheNamespace": "aefd64f8", 
        "type": "AWS_PROXY"
    }
}

MarkBook-Pro:fixtures marksteele$ curl -v http://localhost:4567/restapis/4u6mm1uzzs/prod/_user_request_/HowMuchIsTheFish
*   Trying ::1...
* TCP_NODELAY set
* Connected to localhost (::1) port 4567 (#0)
> GET /restapis/4u6mm1uzzs/prod/_user_request_/HowMuchIsTheFish HTTP/1.1
> Host: localhost:4567
> User-Agent: curl/7.64.1
> Accept: */*
> 
< HTTP/1.1 201 Created
< Server: BaseHTTP/0.6 Python/3.8.2
< Date: Sun, 07 Jun 2020 21:13:31 GMT
< X-Custom-Header: ASDF
< Content-Length: 25
< Access-Control-Allow-Origin: *
< Access-Control-Allow-Methods: HEAD,GET,PUT,POST,DELETE,OPTIONS,PATCH
< Access-Control-Allow-Headers: authorization,content-type,content-md5,cache-control,x-amz-content-sha256,x-amz-date,x-amz-security-token,x-amz-user-agent,x-amz-target,x-amz-acl,x-amz-version-id,x-localstack-target,x-amz-tagging
< Access-Control-Expose-Headers: x-amz-version-id
< 
* Connection #0 to host localhost left intact
{"message":"Hello World"}* Closing connection 0
MarkBook-Pro:fixtures marksteele$ 
```

Once again, I added debug logging:

```

def is_api_key_valid(is_api_key_required, headers, stage):
    LOGGER.info('Checking API key')
    if not is_api_key_required:
        LOGGER.info('API key is not required')
        return True

    LOGGER.info('API key is required')
    api_key = headers.get('X-API-Key')
    if not api_key:
        LOGGER.info('API key not present')
        return False

    LOGGER.info('Checking if API key is valid')
    return validate_api_key(api_key, stage)

```

Logs are showing:

```
localstack_1  | 2020-06-07T21:13:13:INFO:localstack.services.awslambda.lambda_executors: Checking if there are idle containers.
localstack_1  | 2020-06-07T21:13:31:INFO:localstack.services.apigateway.apigateway_listener: Checking API key
localstack_1  | 2020-06-07T21:13:31:INFO:localstack.services.apigateway.apigateway_listener: API key is not required

```

So again, looks like the code to detect if an api key is required isn't working as expected.

**Comment by whummer** [comment]
Thanks for the update @marksteele . Still trying to understand and reproduce this issue. In the `get-method` call above, can you please also post the details of the corresponding `resource`? As we're ultimately calling the `/HowMuchIsTheFish` endpoint, would like to understand which resource this path corresponds to, and how it is configured. Thanks for helping debug this issue.

**Comment by marksteele** [comment]
I literally cannot possibly provide more context. Please re-read my submission. If you simply follow the instructions, you will reproduce the issue. At the end of the bash script, it calls the endpoint without an API key and gets a response (it shouldn't).

If you actually try testing this yourself you should come to the same conclusion (create a lambda, create a rest api, methods, resources, integrations, etc...)

**Comment by whummer** [comment]
Thanks for the update @marksteele . Indeed, we've now been able to reproduce this issue - thanks again for providing the comprehensive example! This should be fixed in #2785 - when you get a chance, can you please give it another try with the `latest` version of the Docker image? Please let us know if there are still any other issues related to API keys. Thanks for your help.

**Comment by marksteele** [comment]
I ran my script against latest, still same issue. Have you guys actually tried to run the script to test your fixes?
